lib/utils/io.py

```diff
@@ -230,17 +230,15 @@ def WriteFile(file_name, fn=None, data=None,
       if callable(prewrite):
         prewrite(fd)
       if data is not None:
-        if isinstance(data, unicode):
-          data = data.encode()
-        assert isinstance(data, str)
-        to_write = len(data)
-        offset = 0
-        while offset < to_write:
-          written = os.write(fd, buffer(data, offset))
-          assert written >= 0
-          assert written <= to_write - offset
-          offset += written
-        assert offset == to_write
+        assert isinstance(data, (str, bytes))
+        if isinstance(data, str):
+          open_mode = "w"
+          encoding = "utf-8"
+        else:
+          open_mode = "wb"
+          encoding = None
+        with open(fd, open_mode, closefd=False, encoding=encoding) as out:
+          out.write(data)
```

Comment:
Yes, it does. open() returns a subclass of BufferedIOBase. For text mode, it uses line buffering by default. Now that I read the original code again, it's obvious that we wanted to do zero-copy writes, which is the opposite from using buffered/batched writes. Do you think it still makes sense? I think that switching this over to block buffering with e.g. an 8kB-long buffer shouldn't do much harm.

Reply:
I took a quick look at callers of WriteFile. Most of this data is small, so trying to do no-buffering doesn't matter much.

Also, this was doing no-buffering user-side, kernel side it was still a normal write going through the page-cache and not an `O_DIRECT` write, so I think this was rather trying to do something else - avoid partial writes caused by signals. That seems to me however a rare occurence, but who knows (how things behaved >10 years ago). In any case, [BufferedIOBase](https://docs.python.org/3.7/library/io.html#io.BufferedIOBase.write) says whole input is written, always, or an exception is raised, so this should go away indeed.

Note that below we do an os.fsync(fd) - I think we need first to call `fd.sync()` here to ensure the (Python/userspace) buffers are sent to the kernel before attempting to flush the kernel buffer.

Reply:
> Note that below we do an os.fsync(fd) - I think we need first to call `fd.sync()` here to ensure the (Python/userspace) buffers are sent to the kernel before attempting to flush the kernel buffer.

I think that's not necessary, because BufferedIOReader flushes the buffer on close, and we're using it as a context manager which calls close() when done.

Reply:
Ah, right. To be honest the fd-vs-out distiction here is a bit confusing; but now the `closefd=False` makes sense :)